### PR TITLE
Inbox/boxes

### PR DESCRIPTION
**Pull request by NelsonVides**
The idea of this complex PR is many-fold. The first bonus here is that there's a lot of simplifying function clauses to make code more readable, which makes the diff a bit more dirty. The rationale behind that is that the usual flow was taking a row from the DB as a big tuple where the order matters, then parsing everything and turning it into a map where fields have names, and then matching those names in the next function call and passing a ton of parameters around a long call-stack. This is unreadable and makes very long lines. Now, we pass the entire map all the way through the call-stack and on each function clause we extract exactly the field we need.

But the real value of this PR is the boxes. Unfortunately, the previous iteration of inbox changes created the archive as a boolean flag, which, having only two possible values, was too inflexible. The idea now is to be able to classify entries into many boxes, so that for example a client can build a UI a-la email, where you can throw conversations into a spam folder, a favourites one, or other maybe privileged folder on some corporate environment that an admin sets for their users.

There's one point where I plan to use this in the future: together with the work I've done on asynchronous inbox: currently, it's dangerously inconsistent. If on two nodes one writes new messages to a user's conversation with a room, and in another node that user is removed from the room, the order in which those events are flushed into the DB is now not ensured, and the risk of the removal being flushed before the insert, and therefore the insert recovering a conversation forever, is too big. With the concept of boxes, the async backend can have a `deleted` box, so that removals aren't really removals, but just moves to another box, which are atomic and can interleave with modifications to that entry's other columns. Then asynchronously the `deleted` boxes can be collected, much like email clears the Bin after so many days.

Note that one part I regret about this PR is that it makes the inbox migration a bit more heavy on live and fat systems, as there's now a lot of rows that require new sizing, padding, and migration of its contents. But well, it's the best balance I could find.

### CODE REVIEW

**Comment by mongoose-im** [comment]
[dynamic_domains_mssql_mnesia_24](https://circleci.com/gh/esl/MongooseIM/116546) / odbc_mssql_mnesia / 523ab2873c4e48c2b503b890f9c30c989ee04979
Reports [root](http://esl.github.io/circleci-mim-results/s3_reports.html?prefix=PR/3608/116546/odbc_mssql_mnesia.24.0.5)

**Comment by mongoose-im** [comment]
[small_tests_24](https://circleci.com/gh/esl/MongooseIM/116568) / small_tests / 880e366dc7830fc6bb685122c62ec79fb7311f95
Reports [root](http://esl.github.io/circleci-mim-results/s3_reports.html?prefix=PR/3608/116568/small_tests.24.0.5) / [small](https://circleci-mim-results.s3.eu-central-1.amazonaws.com/PR/3608/116568/small_tests.24.0.5/small/ct_run.mongooseim@localhost.2022-03-23_19.10.29/lib.mongooseim.logs/run.2022-03-23_19.10.29/suite.log.html)

**Comment by mongoose-im** [comment]
[small_tests_24](https://circleci.com/gh/esl/MongooseIM/116594) / small_tests / 6a37c79e83225d975807cea458e8d1c387dd9a69
Reports [root](http://esl.github.io/circleci-mim-results/s3_reports.html?prefix=PR/3608/116594/small_tests.24.0.5) / [small](https://circleci-mim-results.s3.eu-central-1.amazonaws.com/PR/3608/116594/small_tests.24.0.5/small/ct_run.mongooseim@localhost.2022-03-23_20.27.27/lib.mongooseim.logs/run.2022-03-23_20.27.28/suite.log.html)

File: test/config_parser_SUITE.erl [hunk 1]
```diff
@@ -1733,12 +1733,16 @@ mod_inbox(_Config) ->
     ?cfgh(P ++ [async_writer], #{pool_size => 8}, T(#{<<"async_writer">> => #{<<"pool_size">> => 8}})),
     ?cfgh(P ++ [reset_markers], ChatMarkers, T(#{<<"reset_markers">> => ChatMarkers})),
     ?cfgh(P ++ [groupchat], [muc, muclight], T(#{<<"groupchat">> => [<<"muc">>, <<"muclight">>]})),
+    ?cfgh(P ++ [boxes],
+          [<<"inbox">>, <<"archive">>, <<"favourites">>, <<"spam">>],
+          T(#{<<"boxes">> => [<<"favourites">>, <<"spam">>]})),
```
Comment (chrzaszcz) [inline test/config_parser_SUITE.erl:1738]: If we decide to check for duplicates, this is the place to do so.

**Review by chrzaszcz — COMMENTED** [review]
It all looks good in general, I commented on some details.

File: big_tests/tests/inbox_extensions_SUITE.erl [hunk 2]
```diff
@@ -258,6 +281,142 @@ read_unread_entry_with_two_messages_when_set_unread_then_unread_count_stays_in_t
...
+        inbox_helper:check_inbox(Bob, [#conv{unread = 1, from = Alice, to = Bob, content = Body}]),
+        set_inbox_properties(Bob, Alice, [{box, other}]),
+        inbox_helper:check_inbox(Bob, [#conv{unread = 1, from = Alice, to = Bob, content = Body}],
+                                 #{box => other}),
+        % Then bob decides to recover the conversation
+        set_inbox_properties(Bob, Alice, [{box, inbox}]),
+        % Then the conversation is in the inbox and not in the archive box
+        inbox_helper:check_inbox(Bob, [#conv{unread = 1, from = Alice, to = Bob, content = Body}],
+                                 #{box => inbox}),
+        inbox_helper:check_inbox(Bob, [], #{box => archive})
```
[the remark this answers is not in the document]
Comment (NelsonVides) [reply big_tests/tests/inbox_extensions_SUITE.erl:417]: It's just checking that it is not in the archive box to be precise, note the second argument being an empty list :)

File: src/inbox/mod_inbox.erl [hunk 3]
```diff
@@ -142,22 +146,26 @@ async_config_spec() ->
        include = always
       }.
 
+process_inbox_boxes(Config = #{boxes := Boxes}) ->
+    AllBoxes = [<<"inbox">>, <<"archive">> | Boxes ],
```
[the remark this answers is not in the document]
Comment (NelsonVides) [reply src/inbox/mod_inbox.erl:150]: Very good point. Likewise with duplicates, cleaning those things promptly.

**Review by chrzaszcz — APPROVED** [review]
Looks good 👍

**Pull request merged**